### DISCUSSION TITLE
fast test usenet server ... spit out the same 512.000 bytes articles each time

### DISCUSSION THREAD

**Discussion by sanderjo** [sabnzbd/sabnzbd#2400] [Ideas]
@Safihre @puzzledsab  @mnightingale @animetosho @thezoggy 

Do we still need/want a fast test usenet server, without any disk access server side?

If so, based on nzbget's method, I think we can quite easily: let the server spit out the same article each time, only setting the first two lines uniquely. And those lines are specified by the NZB & client in the NNTP session. 
No cleverness nor memory nor files nor cache files needed on the server side. So the server will not be a notable bottleneck.

I think this can be the fastest test usenet we can create.

I think I can make a prototype in python. Hopefully someone else can write it in C or Go or JS?

Usage
```
./fastest_test_usenet_server --create_blocks 1000 # just to create NZB with resulting size of 1000*512000 = 512 MiB
./fastest_test_usenet_server --create_blocks 20000 # just to create NZB with resulting  size of 20000*512000 = 10GiB

./fastest_test_usenet_server # serve anything the client asks for

```
Or, taking into my remark at the bottom of this post:

```
./fastest_test_usenet_server --blocks 2000 # of 
... Creating NZB "my_big_file_1024MB.bin"
... Serving 50 threads
```

For the example below, we take 2000 segments of 512000 byte each = 1024000000, so 1GB.

```
fallocate -l 1024000000 my_big_file.bin

1024000000 jan  2 21:55 my_big_file.bin
```

`nzbget --nserv -p 9999 -d data_nzbget/ -c cache_data_nzbget/  -z 512000`

will create an NZB with contents:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE nzb PUBLIC "-//newzBin//DTD NZB 1.0//EN" "http://www.newzbin.com/DTD/nzb/nzb-1.0.dtd">
<nzb xmlns="http://www.newzbin.com/DTD/2003/nzb">
<file poster="nserv" date="1672693388" subject="&quot;my_big_file.bin&quot; yEnc (1/2000)">
<groups>
<group>alt.binaries.test</group>
</groups>
<segments>
<segment bytes="512000" number="1">my_big_file.bin?1=0:512000</segment>
<segment bytes="512000" number="2">my_big_file.bin?2=512000:512000</segment>
<segment bytes="512000" number="3">my_big_file.bin?3=1024000:512000</segment>
<segment bytes="512000" number="4">my_big_file.bin?4=1536000:512000</segment>
<segment bytes="512000" number="5">my_big_file.bin?5=2048000:512000</segment>
<segment bytes="512000" number="6">my_big_file.bin?6=2560000:512000</segment>
<segment bytes="512000" number="7">my_big_file.bin?7=3072000:512000</segment>
...
<segment bytes="512000" number="222">my_big_file.bin?222=113152000:512000</segment>
...
<segment bytes="512000" number="1994">my_big_file.bin?1994=1020416000:512000</segment>
<segment bytes="512000" number="1995">my_big_file.bin?1995=1020928000:512000</segment>
<segment bytes="512000" number="1996">my_big_file.bin?1996=1021440000:512000</segment>
<segment bytes="512000" number="1997">my_big_file.bin?1997=1021952000:512000</segment>
<segment bytes="512000" number="1998">my_big_file.bin?1998=1022464000:512000</segment>
<segment bytes="512000" number="1999">my_big_file.bin?1999=1022976000:512000</segment>
<segment bytes="512000" number="2000">my_big_file.bin?2000=1023488000:512000</segment>
</segments>
</file>
</nzb>

```

So, the clever thing about that NZB: based on the NZB, the NZB client (like SAB) will tell nzbget nserv from which file which parts it wants: `my_big_file.bin?222=113152000:512000` so "from your file my_big_file_bin, starting from 113152000 give me 512000 bytes"

Let's look at such a artile: two starting lines, then all "*" lines: chr(0) + 42 from yencode is "*", and then a closing line:

```
$ head 222\=113152000-512000 
=ybegin part=222 line=128 size=1024000000 name=my_big_file.bin
=ypart begin=113152001 end=113664000
********************************************************************************************************************************
********************************************************************************************************************************
********************************************************************************************************************************
********************************************************************************************************************************
********************************************************************************************************************************
********************************************************************************************************************************
********************************************************************************************************************************
********************************************************************************************************************************

$ tail 222\=113152000-512000 
********************************************************************************************************************************
********************************************************************************************************************************
********************************************************************************************************************************
********************************************************************************************************************************
********************************************************************************************************************************
********************************************************************************************************************************
********************************************************************************************************************************
********************************************************************************************************************************
********************************************************************************************************************************
=yend size=512000 part=0 pcrc32=54c9a707

```
So quite easy: 4000 lines of with 128 times "*" plus \r\n, with 2 lines at the start, 1 line at the end.

Comparing two cached articles shows only the first two lines are unique, with the uniqe numbers from the NZB file, as thus as specified by the client `my_big_file.bin?222=113152000:512000`

```
$ diff 222\=113152000-512000 223\=113664000-512000 
1,2c1,2
< =ybegin part=222 line=128 size=1024000000 name=my_big_file.bin
< =ypart begin=113152001 end=113664000
---
> =ybegin part=223 line=128 size=1024000000 name=my_big_file.bin
> =ypart begin=113664001 end=114176000

```
Again: no difference in all lines after line 1 and 2. 

The only thing I wonder about: `size=1024000000` ... that is not in the NZB ... so the client can't tell that to the server. So if SAB needs that (?), specify that on the server side?

**Comment by mnightingale** [comment]
I ended up creating my own in Go, I haven't finished with it but the current version is available [here](https://gist.github.com/mnightingale/7dd2cab7b219e49afcb7ec78df7f15ac).
I found NServ was hitting a bottleneck at around 1 GB/s, I'm not sure if it's because it still has to read the articles from from somewhere even though I had it setup reading from a RAM disk - I found the in-memory option it has unstable.

Outside of Python from a Go client I could read 10GB at >6.5 GB/s

@animetosho shared their [dummy NNTP server](https://gist.github.com/animetosho/e1f26ea6f213f8808d1cabebbb91460e) which has the nice touch that it generates random data for the article and handles a few other commands - I currently just return a zero filled binary file with a hardcoded CRC.

**Comment by puzzledsab** [comment]
I hacked together a script for generating nzbs for the dummy server here: https://gist.github.com/puzzledsab/b71769d478207e5220cc31f01165eb64

Unfortunately I can't test the nzb it generates right now so sorry if it doesn't work.
`perl nzbgen.pl <number of files> > testfile.nzb`

I assume the server can be run using nodejs. I've never done server side JS programming.
